cranelift/codegen/src/isa/aarch64/lower.isle

```diff
@@ -695,3 +695,222 @@
       (value_reg (alu_rs_imm_logic (ALUOp.EorNot64) $I64 x y)))
 
 (rule (lower (has_type $I128 (bxor_not x y))) (i128_alu_bitop (ALUOp.EorNot64) x y))
+
+;;;; Rules for `ishl` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; Shift for i8/i16/i32.
+(rule (lower (has_type (fits_in_32 ty) (ishl x y)))
+      (value_reg (do_shift (ALUOp.Lsl32) ty (put_in_reg x) y)))
+
+;; Shift for i64.
+(rule (lower (has_type $I64 (ishl x y)))
+      (value_reg (do_shift (ALUOp.Lsl64) $I64 (put_in_reg x) y)))
+
+;; Shift for i128.
+(rule (lower (has_type $I128 (ishl x y)))
+      (lower_shl128 (put_in_regs x) (value_regs_get (put_in_regs y) 0)))
+
+;;     lsl     tmp3, src_lo, amt
+;;     lsl     tmp2, src_hi, amt
+;;     mvn     inv_amt, amt
+;;     lsr     tmp1, src_lo, #1
+;;     lsr     tmp1, tmp1, inv_amt
+;;     tst     amt, #0x40
+;;     orr     tmp2, tmp2, tmp1
+;;     csel    dst_hi, tmp3, tmp2, ne
+;;     csel    dst_lo, xzr, tmp3, ne
+(decl lower_shl128 (ValueRegs Reg) ValueRegs)
+(rule (lower_shl128 src amt)
+      (let (
+          (src_lo Reg (value_regs_get src 0))
+          (src_hi Reg (value_regs_get src 1))
+          (lo_lshift Reg (alu_rrr (ALUOp.Lsl64) src_lo amt))
```

Comment:
This is definitely a pre-existing thing, but reading through this code it's becoming pretty apparent to me that having helpers to be able to write e.g. `(lsl64 src lo_amt)` rather than `(alu_rrr (ALUOp.Lsl64) ...)` would make things a lot more readable. I know `ALUOp` can be used in multiple instruction formats so we have duplication (lsl64 with immediate vs not, add64 with immediate, extend, shift, ...) but it seems reasonable to me to pay that cost in order to have super-clear lowering patterns. Thoughts? (This is out-of-scope for this PR of course, would be a followup refactor if we do it)

Reply:
I originally started to do that but the idioms of the AArch64 backend are what stopped me. On x86 each instruction seems to largely be used with one variant of the `Inst` enum, but on AArch64 one `ALUOp` can be used with multiple variants (as you've pointed out). 

Basically I wasn't sure how to resolve that. I figured that going down the route of `alu_rrr` inst builders was not horrible but I agree it's not exactly the most readable. I'd be happy to change though if you've got an idea of how to name the constructors or resolve this

Reply:
Maybe something like: just the mnemonic for the all-regs case (so `(add64 rn rm)`, `(lsl64 rn rm)`, etc.), and then e.g. `add64_imm12`, `add64_shift`, `add64_extend`? This also fixes the issue where illegal combinations are allowed currently: e.g. right now we can put the pieces together to make an `add64` with `ImmLogic` but that's disallowed (should hit a panic during emission) -- better to catch statically.

Not super-urgent, just better to do sooner than later if we're going to do it!